<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2021-02-09" />

<title>Reproducible scripting, AlphaSimR, Homework</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">PB7170</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Reproducible scripting, AlphaSimR, Homework</h1>
<h4 class="author">Jean-Luc Jannink</h4>
<address class="author_afil">
USDA-ARS / Cornell<br><h4 class="date">February 9, 2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-02-15
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>PB7170/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210208code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210208)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210208code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210208)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong124bcdf"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> 124bcdf </a>
</p>
</div>
<div id="strongRepositoryversionstrong124bcdf" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version 124bcdf. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    analysis/.Rhistory

Untracked files:
    Untracked:  CentralLimitTheorem.Rmd
    Untracked:  output/QTLgenotypes.csv
    Untracked:  workflowr_update.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Lab2_ProgrammingAlphaSimR.Rmd</code>) and HTML (<code>docs/Lab2_ProgrammingAlphaSimR.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
124bcdf
</td>
<td>
plantwizard
</td>
<td>
2021-02-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
0b0f372
</td>
<td>
plantwizard
</td>
<td>
2021-02-15
</td>
<td>
Update new files
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>
<div id="preliminaries" class="section level1">
<h1>Preliminaries</h1>
<div id="learning-objectives" class="section level3">
<h3>Learning objectives</h3>
<ol style="list-style-type: decimal">
<li>A bit more on reproducible scripting<br />
Some other things that matter to me to ensure communicability and reproducibility</li>
</ol>
<ol style="list-style-type: lower-roman">
<li>The main README.md file should contain information about the conditions under-which the script was run<br />
</li>
<li>I like the script to have a predictable strucure: load and document the packages used, if there is a stochastic component, set the random seed, set and document the parameter values that will affect the script output<br />
</li>
<li>Use a consistent directory structure. Happily, such a consistent structure is enforced by <code>workflowr</code><br />
</li>
<li>Use the package <code>here</code> to make file locations relative to the base project folder</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li><p>AlphaSimR<br />
<code>AlphaSimR</code> is a package to simulate breeding populations and tasks. It is not completely intuitive nor completely well-documented. We will want to use it ultimately to optimize breeding schemes. Hopefully, we will be able to contrast it’s output with the output of deterministic analytic optimization</p></li>
<li><p>Prepare a Homework<br />
In class, we have discussed two interesting phenomena related to quantitative traits:<br />
</p></li>
</ol>
<ol style="list-style-type: lower-roman">
<li>The “outbreak of variation” that occurs when a heterozygous individual is self-fertilized<br />
</li>
<li>Regression toward the mean between parents and progeny<br />
This Rmarkdown script contains code to illustrate the former. Your homework will be to write a script that illustrates the latter</li>
</ol>
</div>
</div>
<div id="ordering-of-the-script" class="section level1">
<h1>Ordering of the script</h1>
<p>It’s good to have all scripts in the same order with respect to standard tasks</p>
<div id="load-packages-first" class="section level3">
<h3>Load packages first</h3>
<p>If your script depends on external packages, load them at the beginning. This shows users early on what the script dependencies are.</p>
<pre class="r"><code>packages_used &lt;- c(&quot;AlphaSimR&quot;, &quot;tidyverse&quot;, &quot;workflowr&quot;, &quot;here&quot;)
ip &lt;- installed.packages()
for (package in packages_used){
  if (!(package %in% ip[,&quot;Package&quot;])) install.packages(package)
}#END packages_used</code></pre>
<p>Notice the conflicts report from loading tidyverse. Two packages (<code>dplyr</code> and <code>stats</code>) both have a function called <code>filter</code>. Since dplyr was loaded <em>after</em> stats, if you use the function <code>filter</code>, it will go to the dplyr version. It is dangerous to rely on what order packages have been loaded to determine which <code>filter</code> function you get. R syntax to prevent ambiguity is to write either <code>dplyr::filter</code> or <code>stats::filter</code>. Using that syntax will make your code more reproducible.</p>
</div>
<div id="set-file-locations-relative-to-the-project-folder-using-here" class="section level3">
<h3>Set file locations relative to the project folder using <code>here</code></h3>
<pre class="r"><code>here::i_am(&quot;analysis/Lab2_ProgrammingAlphaSimR.Rmd&quot;)</code></pre>
</div>
<div id="document-packages-used" class="section level3">
<h3>Document packages used</h3>
<p>This chunk creates a “README.md” file that documents the packages and versions used for future reference.</p>
<pre class="r"><code>source(here::here(&quot;code/addToREADME.R&quot;))
addToREADME(paste0(&quot;## &quot;, rmarkdown::metadata$title), append=F)
addToREADME(c(date(), &quot;&quot;))

packages_info &lt;- ip[packages_used, c(&quot;Package&quot;, &quot;Version&quot;, &quot;Built&quot;)]
addToREADME(c(&quot;The packages used in this script are:&quot;, &quot;Package, Version, Built&quot;))
apply(packages_info, 1, function(vec) addToREADME(paste(vec, collapse=&quot; &quot;)))</code></pre>
<pre><code>               AlphaSimR                tidyverse                workflowr 
&quot;AlphaSimR 0.13.0 4.0.3&quot;  &quot;tidyverse 1.3.0 4.0.3&quot;  &quot;workflowr 1.6.2 4.0.3&quot; 
                    here 
      &quot;here 1.0.1 4.0.3&quot; </code></pre>
<pre class="r"><code>addToREADME(&quot;&quot;)</code></pre>
<p>Hyperlink the <a href="README.md">README file</a> to the report. That makes it easy to find.</p>
</div>
<div id="set-random-seed" class="section level3">
<h3>Set random seed</h3>
<p>AlphaSimR generates many random numbers (e.g., to simulate Mendelian random segregation). If you want the result of the analysis to come out the same each time (there are pros and cons) you need to set the random seed. Note that <code>workflowr</code> does this by default. If you are not using that package, then be explicit.</p>
<pre class="r"><code>random_seed &lt;- 45678
set.seed(random_seed)
addToREADME(c(paste(&quot;The random seed is&quot;, random_seed), &quot;&quot;))</code></pre>
</div>
<div id="script-parameters" class="section level3">
<h3>Script parameters</h3>
<p>If the behavior of your script depends on parameters that you set, initialize them early on.</p>
<pre class="r"><code>nFounders &lt;- 100
nChr &lt;- 10 # Number of chromosomes
nSitesPerChr &lt;- 1000 # Number of segregating sites _per chromosome_
nQTLperChr &lt;- 10 # Vary this parameter to get oligo- versus poly- genic traits</code></pre>
</div>
<div id="parameters-to-readme" class="section level3">
<h3>Parameters to README</h3>
<p>It’s good to have all the information together in the README. Chunks of code like this do not need to be included in the report. To exclude them use the “include=FALSE” option in the chunk.</p>
</div>
</div>
<div id="simulating-some-classical-results" class="section level1">
<h1>Simulating some classical results</h1>
<p>This script uses <code>AlphaSimR</code> to simulate the “outbreak of variation” that arises when you self-fertilize a hybrid.</p>
<div id="alphasimr-populations" class="section level3">
<h3>AlphaSimR populations</h3>
<p>The basic object of <code>AlphaSimR</code> is the population. To make founders, you first make founder haplotypes from a coalescent simulation, then you define simulation parameters that will link their genetic variation to phenotypic variation, then you make a first diploid population from the founder haplotypes.</p>
<pre class="r"><code># Create haplotypes for founder population of outbred individuals
# Note: default effective population size for runMacs is 100
founderHaps &lt;- AlphaSimR::runMacs(nInd=nFounders, nChr=nChr, segSites=nSitesPerChr)

# New global simulation parameters from founder haplotypes
SP &lt;- AlphaSimR::SimParam$new(founderHaps)
# Additive trait architecture
# By default, the genetic variance will be 1
SP$addTraitA(nQtlPerChr=nQTLperChr)

# Create a new population of founders
founders &lt;- AlphaSimR::newPop(founderHaps, simParam=SP)
str(founders)</code></pre>
<pre><code>Formal class &#39;Pop&#39; [package &quot;AlphaSimR&quot;] with 17 slots
  ..@ id     : chr [1:100] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
  ..@ mother : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ father : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ sex    : chr [1:100] &quot;H&quot; &quot;H&quot; &quot;H&quot; &quot;H&quot; ...
  ..@ nTraits: int 1
  ..@ gv     : num [1:100, 1] 0.0946 -0.4174 -2.3407 -0.8747 -1.8342 ...
  ..@ pheno  : num [1:100, 1] NA NA NA NA NA NA NA NA NA NA ...
  ..@ ebv    : num[1:100, 0 ] 
  ..@ gxe    :List of 1
  .. ..$ : NULL
  ..@ fixEff : int [1:100] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ reps   : num [1:100] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ misc   :List of 100
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
  .. .. [list output truncated]
  ..@ nInd   : int 100
  ..@ nChr   : int 10
  ..@ ploidy : int 2
  ..@ nLoci  : int [1:10] 1000 1000 1000 1000 1000 1000 1000 1000 1000 1000
  ..@ geno   :List of 10
  .. ..$ : raw [1:125, 1:2, 1:100] 88 47 6c 01 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 9a 5b 22 1b ...
  .. ..$ : raw [1:125, 1:2, 1:100] 44 07 02 4e ...
  .. ..$ : raw [1:125, 1:2, 1:100] 01 45 70 90 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 3f a1 41 66 ...
  .. ..$ : raw [1:125, 1:2, 1:100] ee 1f 15 21 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 08 09 e2 5e ...
  .. ..$ : raw [1:125, 1:2, 1:100] 0c 18 81 17 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 48 40 3d 02 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 46 18 08 32 ...
  .. ..- attr(*, &quot;dim&quot;)= int [1:2] 10 1</code></pre>
</div>
<div id="population-information" class="section level3">
<h3>Population information</h3>
<p>The population has <code>id</code>s. The <code>@mother</code> and <code>@father</code> ids are all zero because this population was made from founder haplotypes, and so does not have diploid parents. The genotypic values <code>gv</code> of the population are calculated for the trait created using <code>SP$addTraitA(nQtlPerChr=nQTLperChr)</code>. Given that there are 10 chromosomes and 10 QTL per chromosome, there are <code>3^(nChr*nQTLperChr)</code> = 5.153775210^{47} different possible genotypic values. These genotypic values are accessible with the function <code>gv(founders)</code></p>
<p>From here, you can treat this population like a named vector using the square braces extraction operator <code>[ ]</code>. Extract individuals by their <code>@id</code> or just by their order in the population using an integer index. For example, pick three random individuals from a population and list their ids. Pick the one with the first id in alphabetical order.</p>
<pre class="r"><code>test &lt;- founders[sample(nFounders, 3)]
testID &lt;- test@id
firstInd &lt;- test[order(testID)[1]]
print(testID)</code></pre>
<pre><code>[1] &quot;76&quot; &quot;84&quot; &quot;16&quot;</code></pre>
<pre class="r"><code>print(firstInd)</code></pre>
<pre><code>An object of class &quot;Pop&quot; 
Ploidy: 2 
Individuals: 1 
Chromosomes: 10 
Loci: 10000 
Traits: 1 </code></pre>
<pre class="r"><code>print(firstInd@id)</code></pre>
<pre><code>[1] &quot;16&quot;</code></pre>
</div>
<div id="outbreak-of-variation" class="section level3">
<h3>Outbreak of variation</h3>
<p>Emerson and East (1913) showed that if you crossed to inbreds, the hybrid had similar variation to each inbred, but if you then selfed the hybrid, the offspring varied substantially. This code simulates that result. First, self the founders to homozygosity. The function <code>self</code> self-fertilizes individuals from the population. By default, it creates one selfed individual per parent (controllable with the parameter <code>nProgeny</code>), so this works nicely for single-seed descent.</p>
<pre class="r"><code># Self-pollinate to for a few generations
nGenSelf &lt;- 3
inbredPop &lt;- founders
for (gen in 1:nGenSelf){
  inbredPop &lt;- AlphaSimR::self(inbredPop)
}</code></pre>
</div>
<div id="check-homozygosity" class="section level3">
<h3>Check homozygosity</h3>
<p>Just a sanity check that this has, in fact, created a population of 100 individuals that are appropriately homozygous. Loci are coded 0, 1, 2. So <code>qtl == 1</code> represents the case were a locus is heterozygous. <code>sum(qtl == 1)</code> counts those cases.</p>
<pre class="r"><code>qtl &lt;- AlphaSimR::pullQtlGeno(inbredPop)
if (nrow(qtl) != nFounders) stop(&quot;The number of individuals is unexpected&quot;)
if (ncol(qtl) != nChr * nQTLperChr) stop(&quot;The number of QTL is unexpected&quot;)
fracHet &lt;- sum(qtl == 1) / (nFounders * nChr * nQTLperChr)
cat(&quot;Expected fraction heterozygous&quot;, 1 / 2^nGenSelf, &quot;\n&quot;,
    &quot;Observed fraction heterozygous&quot;, fracHet, &quot;+/-&quot;, 
    round(2*sqrt(fracHet*(1-fracHet)/(nFounders*nChr*nQTLperChr)), 3), &quot;\n&quot;)</code></pre>
<pre><code>Expected fraction heterozygous 0.125 
 Observed fraction heterozygous 0.0292 +/- 0.003 </code></pre>
<blockquote>
<p>What was wrong with my reasoning about the Expected fraction heterozygous?</p>
</blockquote>
</div>
<div id="simulate-outbreak-of-variation" class="section level3">
<h3>Simulate outbreak of variation</h3>
<p>We will pick the pair of most average or most extreme inbred individuals, in terms of their genetic value, cross those pairs, then self-fertilize them to test their variation. We will assume a trait that has a heritability of 0.5 in the base, non-inbred population population. In <code>AlphaSimR</code>, that means genetic and error variances of 1. Find the average pair by ordering and picking two in the middle. Find the extreme pair by picking the first and last.</p>
<pre class="r"><code>indGVs &lt;- AlphaSimR::gv(inbredPop)
averageInd &lt;- inbredPop[order(indGVs)[50:51]]
extremeInd &lt;- inbredPop[order(indGVs)[c(1, nFounders)]]</code></pre>
</div>
<div id="cross-the-pair-to-make-f1-hybrid" class="section level3">
<h3>Cross the pair to make F1 hybrid</h3>
<p>The crossPlan is a <strong>two-column matrix</strong> with as many rows as the number of crosses you want to make, the first column is the <code>@id</code> or the index of the seed parent, and likewise for the pollen parent in the second column. To make an F1, you just want to make one cross, so the matrix is one row. You just want to cross individual 1 with individual 2, so the row contains 1:2.</p>
<pre class="r"><code>crossPlan &lt;- matrix(1:2, nrow=1, ncol=2)
f1_average &lt;- AlphaSimR::makeCross(averageInd, crossPlan)
f1_extreme &lt;- AlphaSimR::makeCross(extremeInd, crossPlan)</code></pre>
</div>
<div id="make-f2-and-observe-phenotypic-variance" class="section level3">
<h3>Make F2 and observe phenotypic variance</h3>
<p>When you first make a population, <code>AlphaSimR</code> does not assume it has been phenotyped. You can phenotype it using the <code>setPheno</code> function. Note that if you use <code>setPheno</code> on the same population a second time, that will overwrite the phenotypes from the first time. The genotypic variance can be retrieved using the function <code>varG</code>. Really, <code>varG</code> gives all we need to know, but of course that variance is generally not observable in non-simulated reality.</p>
<pre class="r"><code>f2_average &lt;- AlphaSimR::self(f1_average, nProgeny=100)
f2_average &lt;- AlphaSimR::setPheno(f2_average, varE=1)
cat(&quot;Genotypic variance of F2 between average pair&quot;, round(AlphaSimR::varG(f2_average), 3))</code></pre>
<pre><code>Genotypic variance of F2 between average pair 0.455</code></pre>
<pre class="r"><code>hist(AlphaSimR::pheno(f2_average), main=&quot;Histogram of Average-Pair F2 Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab2_ProgrammingAlphaSimR.Rmd/Make%20F2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>f2_extreme &lt;- AlphaSimR::self(f1_extreme, nProgeny=100)
f2_extreme &lt;- AlphaSimR::setPheno(f2_extreme, varE=1)
cat(&quot;Genotypic variance of F2 between extreme pair&quot;, round(AlphaSimR::varG(f2_extreme), 3))</code></pre>
<pre><code>Genotypic variance of F2 between extreme pair 1.157</code></pre>
<pre class="r"><code>hist(AlphaSimR::pheno(f2_extreme), main=&quot;Histogram of Extreme-Pair F2 Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab2_ProgrammingAlphaSimR.Rmd/Make%20F2-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="observe-f1-phenotypic-variance" class="section level3">
<h3>Observe F1 phenotypic variance</h3>
<p>We only made a single F1 plant. <code>AlphaSimR</code> can’t give that plant different phenotypes. So to pretend to be Emerson and East, we have to make, say, 100 F1 individuals. Do this just for the f1_extreme case. Here, <code>varG</code> may or may not be zero, depending on how many generations of selfing you did. Now the <code>crossPlan</code> has many rows indicating the same cross.</p>
<pre class="r"><code>crossPlan &lt;- matrix(1:2, nrow=100, ncol=2, byrow=T)
f1_extremeMany &lt;- AlphaSimR::makeCross(extremeInd, crossPlan)
f1_extremeMany &lt;- AlphaSimR::setPheno(f1_extremeMany, varE=1)
cat(&quot;Genotypic variance of F1 between extreme pair&quot;, round(AlphaSimR::varG(f1_extremeMany), 3))</code></pre>
<pre><code>Genotypic variance of F1 between extreme pair 0.032</code></pre>
<pre class="r"><code>hist(AlphaSimR::pheno(f1_extremeMany), main=&quot;Histogram of Extreme-Pair *F1* Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab2_ProgrammingAlphaSimR.Rmd/Make%20many%20hybrids-1.png" width="672" style="display: block; margin: auto;" /></p>
<blockquote>
<p>What conditions will tend to make the outbreak of variation under the “average” scenario similar to the “extreme” scenario?</p>
</blockquote>
</div>
<div id="illustrate-here-by-writing-a-file" class="section level3">
<h3>Illustrate <code>here</code> by writing a file</h3>
<p>Just to practice again using the here package because it makes life easy.</p>
<pre class="r"><code>utils::write.csv(qtl, here::here(&quot;output&quot;, &quot;QTLgenotypes.csv&quot;))</code></pre>
</div>
</div>
<div id="homework" class="section level1">
<h1>Homework</h1>
<p>Illustrate regression to the mean between parents and offspring using AlphaSimR<br />
1. You know how to create a population – create a population of parents<br />
2. You know how to get phenotypes from that population<br />
3. Assume you want to randomly mate that population to get progeny that will be regressed to the mean. The command for that is <code>AlphaSimR::randCross</code>. Check out its documentation<br />
4. Having made the progeny population, phenotype it also<br />
5. Make a scatterplot of the progeny phenotypes against the parent mean phenotypes<br />
i) This is actually tricky: you are going to have to find the pedigree of the progeny using the <code>@mother</code> and <code>@father</code> ids of the progeny population and match those up to the parent population<br />
ii) Having done that, run the regression of progeny on parent mean<br />
iii) What is the regression coefficient?</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       rstudioapi_0.13  whisker_0.4      knitr_1.30      
 [5] magrittr_2.0.1   workflowr_1.6.2  hms_0.5.3        here_1.0.1      
 [9] R6_2.5.0         rlang_0.4.8      AlphaSimR_0.13.0 stringr_1.4.0   
[13] tools_4.0.3      xfun_0.19        git2r_0.28.0     htmltools_0.5.0 
[17] ellipsis_0.3.1   yaml_2.2.1       digest_0.6.27    rprojroot_2.0.2 
[21] tibble_3.0.4     lifecycle_0.2.0  crayon_1.3.4     readr_1.4.0     
[25] later_1.1.0.1    vctrs_0.3.5      promises_1.1.1   fs_1.5.0        
[29] glue_1.4.2       evaluate_0.14    rmarkdown_2.6    stringi_1.5.3   
[33] compiler_4.0.3   pillar_1.4.7     httpuv_1.5.4     pkgconfig_2.0.3 </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
